File: Form a palindrome - GFG/form-a-palindrome.java
// { Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while(t-- > 0)
        {
            
            String S = read.readLine().trim();
            Solution ob = new Solution();
            System.out.println(ob.findMinInsertions(S));
        }
    }
}// } Driver Code Ends


//User function Template for Java

class Solution{
    private static Integer[][] dp;
    private static int find(int l, int r, String s){
        if(l > r){
            return 0;
        }
        if(dp[l][r] != null){
            return dp[l][r];
        }
        
        if(s.charAt(l) == s.charAt(r)){
            return dp[l][r] = find(l+1, r-1, s);
        }else{
            return dp[l][r] = 1 + Math.min(find(l+1, r, s), find(l, r-1, s));
        }
    }
    
    int findMinInsertions(String S){
        dp = new Integer[S.length()][S.length()];    
        return find(0, S.length()-1, S);    
    }
}

File: GFG/Form a palindrome/Form a palindrome.py
#User function Template for python3

class Solution:
    def countMin(self, Str):
        arr = Str
        brr = arr[::-1]
        
        m = len(arr)
        n = len(brr)
        lcs = [[0]*(n+1) for _ in range(m+1)]
        
        for i in range(1,m+1):
            for j in range(1,n+1):
                if arr[i-1] == brr[j-1]:
                    lcs[i][j] = 1 + lcs[i-1][j-1]
                else:
                    lcs[i][j] = max(lcs[i-1][j], lcs[i][j-1])
        return n - lcs[m][n]

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':

    t = int(input())

    for _ in range(t):
        Str = input()
        

        solObj = Solution()

        print(solObj.countMin(Str))
# } Driver Code Ends
